src/Nethermind/Nethermind.Blockchain/BeaconBlockRoot/BeaconBlockRootHandler.cs

```diff
@@ -1,39 +1,48 @@
 // SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
 // SPDX-License-Identifier: LGPL-3.0-only
 
-using Nethermind.Core.Specs;
+using System;
 using Nethermind.Core;
+using Nethermind.Core.Specs;
+using Nethermind.Crypto;
+using Nethermind.Evm.Tracing;
+using Nethermind.Evm.TransactionProcessing;
 using Nethermind.Int256;
-using Nethermind.State;
-using Nethermind.Core.Crypto;
-using Nethermind.Core.Extensions;
-
-namespace Nethermind.Consensus.BeaconBlockRoot;
 
-public class BeaconBlockRootHandler : IBeaconBlockRootHandler
+namespace Nethermind.Blockchain.BeaconBlockRoot;
+public class BeaconBlockRootHandler(ITransactionProcessor processor) : IBeaconBlockRootHandler
 {
-    public void ApplyContractStateChanges(Block block, IReleaseSpec spec, IWorldState stateProvider)
-    {
-        if (!spec.IsBeaconBlockRootAvailable ||
-            block.IsGenesis ||
-            block.Header.ParentBeaconBlockRoot is null)
-            return;
-
-        Address eip4788Account = spec.Eip4788ContractAddress ?? Eip4788Constants.BeaconRootsAddress;
+    private const long GasLimit = 30_000_000L;
 
-        if (!stateProvider.AccountExists(eip4788Account))
-            return;
-
-        UInt256 timestamp = (UInt256)block.Timestamp;
-        Hash256 parentBeaconBlockRoot = block.ParentBeaconBlockRoot;
-
-        UInt256.Mod(timestamp, Eip4788Constants.HistoryBufferLength, out UInt256 timestampReduced);
-        UInt256 rootIndex = timestampReduced + Eip4788Constants.HistoryBufferLength;
-
-        StorageCell tsStorageCell = new(eip4788Account, timestampReduced);
-        StorageCell brStorageCell = new(eip4788Account, rootIndex);
-
-        stateProvider.Set(tsStorageCell, Bytes.WithoutLeadingZeros(timestamp.ToBigEndian()).ToArray());
-        stateProvider.Set(brStorageCell, Bytes.WithoutLeadingZeros(parentBeaconBlockRoot.Bytes).ToArray());
+    public void StoreBeaconRoot(Block block, IReleaseSpec spec)
+    {
+        BlockHeader? header = block.Header;
+        var canInsertBeaconRoot = spec.IsBeaconBlockRootAvailable
+                                  && !header.IsGenesis
+                                  && header.ParentBeaconBlockRoot is not null;
+
+        if (canInsertBeaconRoot)
+        {
+            Transaction transaction = new()
+            {
+                Value = UInt256.Zero,
+                Data = header.ParentBeaconBlockRoot.Bytes.ToArray(),
+                To = spec.Eip4788ContractAddress ?? Eip4788Constants.BeaconRootsAddress,
+                SenderAddress = Address.SystemUser,
+                GasLimit = GasLimit,
+                GasPrice = UInt256.Zero,
+            };
+
+            transaction.Hash = transaction.CalculateHash();
+
+            try
+            {
+                processor.Execute(transaction, header, NullTxTracer.Instance);
```

Comment:
Do we add Eip4788ContractAddress to access list like geth does? If the call fails should the address stay in the list if so?
Do we need to trace system calls in some additional way?
Is the result the same in case of failure like for other implementations?

Reply:
1. Not sure, but it works on mainnet, so I guess it is the same?
2. Tracing is outside of the scope of this PR. We might want to add special tracing method that will trace system transactions too.
3. Good question, failure should definitely fail the block, not sure about the reason.

Reply:
1. Very theoretical, but it may cost less gas like 29_999_900 and pass but we will charge 30_000_100 and fail. And traces may differ I guess. Traces are important for L2 provers I guess.
2. Yep, I'm not sure at all what to expect for system tx in this regard
3. It should fail silently and proceed with regular txs

Reply:
Geth code: https://github.com/ethereum/go-ethereum/blob/master/core/state_processor.go#L180

Reply:
applied, we need to add tracing in separate PR